integrations/kube-agent-updater/pkg/controller/statefulset.go

```diff
@@ -0,0 +1,284 @@
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type StatefulSetVersionUpdater struct {
+	VersionUpdater
+	kclient.Client
+	Scheme *runtime.Scheme
+}
+
+// Reconcile treats a reconciliation request for a StatefulSet object. It gets the
+// object, retrieves its current version, and calls the VersionUpdater to find
+// if the StatefulSet should be updated. If it's the case, it changes the
+// Teleport image version and updates the StatefulSet in Kubernetes.
+//
+// WARNING: StatefulSets rollouts can end up being stuck because of unready pods.
+// We must delete those unhealthy pods to ensure the rollout is not blocked.
+// Deleting only after an update is not idempotent, and deleting every
+// reconciliation or maintenance might be disruptive and cause misleading error
+// if something else is broken (i.e. the state is invalid but we keep deleting
+// the pods). To mitigate the disruption, we only delete unhealthy pods whose
+// spec is not based on the current PodTemplate.
+func (r *StatefulSetVersionUpdater) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx).WithValues("namespacedname", req.NamespacedName, "kind", "StatefulSet")
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	// Get the object we are reconciling
+	var obj appsv1.StatefulSet
+	if err := r.Get(ctx, req.NamespacedName, &obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, trace.Wrap(err)
+	}
+
+	// Get the current and past version
+	currentVersion, err := getWorkloadVersion(obj.Spec.Template.Spec)
+	if err != nil {
+		switch trace.Unwrap(err).(type) {
+		case *trace.BadParameterError:
+			log.Info("Teleport container found, but failed to get version from the img tag. Will continue and do a version update.")
+		default:
+			return requeueLater, trace.Wrap(err)
+		}
+	}
+
+	// Check if the statefulset is blocked by an old unhealthy pod, if so
+	// attempt to remove it. To ensure idempotentcy this happens also outside
+	// of maintenance windows and if no update is available.
+	// Deferring the unblock allows it to happen after a potential update. We
+	// won't need to wait another reconciliation window, or 10+ minutes for new
+	// pods to be created from the new StatefulSet spec.
+	defer func() {
+		// Note: this might be called often, we might decide to enable cache for pods
+		// for performance purposes (we trade more memory usage on the updater side,
+		// against less latency and calls to the apiserver)
+		if err := r.unblockStatefulSetRollout(ctx, &obj); err != nil {
+			log.Error(err, "statefulset unblocking failed, the rollout might get stuck")
+		}
+	}()
```

Comment:
So this will run on every reconciler tick even if there's no upgrade available it look like. Under what circumstances it can happen that it will actually need to "fix" the SS outside of the upgrade window?

Also, if this needs to run basically on every tick anyway (to clean things up from a failed update I assume), wouldn't making it into a separate controller/service/goroutine be more appropriate (and call this here only after actual update happened)?

Reply:
> if there's no upgrade available it look like. Under what circumstances it can happen that it will actually need to "fix" the SS outside of the upgrade window?

Pods are considered unhealthy after 10 minutes. If a pod is blocking the rollout but not yet unhealthy during the reconciliation, we might update the sts and never delete it. The sts will stay broken until the next maintenance window in 24 hours.

Reply:
I edited the controller to try to unstock pods in the following cases:
- when a maintenance was triggered but no new version was found (unhealthy pods will trigger maintenance)
- when a maintenance was triggered, new version was found but we failed to validate the image
- when the version was successfully updated

We do not try to unblock in the following cases:
- when a maintenance was not triggered (nothing to do, the most common case)
- when we encounter an unknown error when checking maintenance,version,image)
- when we face an error when updating the statefulset (99% chance we conflicted with something else and are being requeued, the update will pass the next time)